### PR TITLE
Enabling labels to be submitted to AI Platform training jobs through a TFX pipeline

### PR DESCRIPTION
**Pull request by axeltidemann**
This pull request is a result of a reported [issue](https://github.com/tensorflow/tfx/issues/4385) highlighting that it is not possible to submit labels with AI Platform training jobs through a TFX pipeline. This is possible on Vertex Pipelines, but a lot of pipelines will still be run on the AI Platform. The design of the solution was discussed with @1025KB, and we agreed on a small change that allows the `Trainer` component to be called with `labels` like this: 

```
tfx.extensions.google_cloud_ai_platform.Trainer(...,
    custom_config = { tfx.extensions.google_cloud_ai_platform.TRAINING_ARGS_KEY: { "scaleTier": ...},
                      "labels": {"team": "awesome" }
    }
)
```

Labels are very useful for billing purposes, which is why this change is desirable.

### CODE REVIEW

**Comment by axeltidemann** [comment]
I have submitted my Contributor License Agreement. Please re-scan.

File: tfx/extensions/google_cloud_ai_platform/training_clients.py [hunk 1]
```diff
@@ -228,6 +230,10 @@ def create_training_job(self, input_dict: Dict[str, List[types.Artifact]],
         {telemetry_utils.LABEL_TFX_EXECUTOR: executor_class_path}):
       job_labels = telemetry_utils.make_labels_dict()
 
+    custom_config = json_utils.loads(
+        exec_properties.get(standard_component_specs.CUSTOM_CONFIG_KEY))
+    job_labels.update(custom_config.get('labels', {}))
```
Comment (1025KB) [inline tfx/extensions/google_cloud_ai_platform/training_clients.py:235]: for cutom_config level keys, you would want to add it like [this](https://github.com/tensorflow/tfx/blob/master/tfx/extensions/google_cloud_ai_platform/trainer/executor.py#L40), thus users can find the keys and use it. consider make it an optional param to pass in this function from executor.py

Comment (axeltidemann) [reply tfx/extensions/google_cloud_ai_platform/training_clients.py]: Good point @1025KB, I have added the `LABELS_KEY` to accommodate this. Minor logic change to merge labels in `executor.py`. In the `CAIPJobClient` I simply pop the `labels` dictionary from the `training_inputs` and update the `job_labels` dictionary.

File: tfx/extensions/google_cloud_ai_platform/trainer/executor.py [hunk 2]
```diff
@@ -97,6 +104,9 @@ def Do(self, input_dict: Dict[str, List[types.Artifact]],
       absl.logging.error(err_msg)
       raise ValueError(err_msg)
 
+    job_args['labels'] = { **job_args.get('labels', {}),
+                           **custom_config.get(LABELS_KEY, {}) }
```
Comment (1025KB) [inline tfx/extensions/google_cloud_ai_platform/trainer/executor.py:108]: Let's pass it in just like job_id (line 110)? or do it like [project](https://github.com/tensorflow/tfx/blob/master/tfx/extensions/google_cloud_ai_platform/runner.py#L191)

as the [job_args](https://github.com/tensorflow/tfx/blob/f84b768f28ed60ab1b51dded73c64718950db1f4/tfx/extensions/google_cloud_ai_platform/training_clients.py#L200) of client.create_training_job is the training_input, so I think it would be good to keep it as it is, and add an optional labels to that function?

Comment (axeltidemann) [reply tfx/extensions/google_cloud_ai_platform/trainer/executor.py]: 1. If it is passed in like `job_id`, then the signatures of `runner.start_cloud_training`, `CAIPJobClient.create_training_job` and `VertexJobClient.create_training_job` would have to change, and it seems a bit convoluted, since `labels` is part of `job_args` for `VertexJobClient` already?
2. Similar argument for doing it like `project`, but it would be slightly simpler than option 1 since you would not have to change `runner.start_cloud_training` in addition.

I see your point about the `job_args` in `CAIPJobClient` should not have `labels`, but it is simply stored there and [popped](https://github.com/axeltidemann/tfx/blob/18f05ff652919f3e8566afd583e4fb860e29b392/tfx/extensions/google_cloud_ai_platform/training_clients.py#L230) before passed on - which is similar to the "stowaway" project you refer to. I personally think my proposal is the least "intrusive" option, but I have implemented option 1 for you to review.

Comment (1025KB) [reply tfx/extensions/google_cloud_ai_platform/trainer/executor.py]: I see your point. if we want to just pass the label in job_args as a workaround for CAIP, then we don't need to add constant to executor.py, those constants are top level keys of custom_config

Comment (axeltidemann) [reply tfx/extensions/google_cloud_ai_platform/trainer/executor.py]: So you mean like my first commit, basically?

Comment (axeltidemann) [reply tfx/extensions/google_cloud_ai_platform/trainer/executor.py]: I will go on holiday leave now, and will check back in two weeks. If I understand you correctly, what you are describing is my first commit - feel free to approve that in the meantime. If not, I am of course happy to continue this when I am back. Happy holidays!

Comment (1025KB) [reply tfx/extensions/google_cloud_ai_platform/trainer/executor.py]: Current code look good (except we want to make labels an optional args for AI platform only)

(Due to Vertex/AI platform diff, I don't think we have a perfect way to align them)

File: tfx/extensions/google_cloud_ai_platform/runner.py [hunk 3]
```diff
@@ -156,6 +156,7 @@ def start_cloud_training(input_dict: Dict[str, List[types.Artifact]],
                          exec_properties: Dict[str, Any],
                          executor_class_path: str,
                          job_args: Dict[str, Any],
+                         job_labels: Dict[str, Any],
```
Comment (1025KB) [inline tfx/extensions/google_cloud_ai_platform/runner.py:159]: can we make job_lables optional (same elsewhere), and comment it's for AI platform only? Thanks!

**Comment by axeltidemann** [comment]
I am working on this, will update as soon as I can.

File: tfx/extensions/google_cloud_ai_platform/trainer/executor.py (hunk 2 above)
Comment (axeltidemann) [reply tfx/extensions/google_cloud_ai_platform/trainer/executor.py]: I have made `job_labels` optional, please review.

**Comment by axeltidemann** [comment]
@1025KB Could you review the changes I made, please?

**Comment by axeltidemann** [comment]
@1025KB Please have a look - it would be awesome to finish this.

**Review by 1025KB — APPROVED** [review]

File: tfx/extensions/google_cloud_ai_platform/runner.py [hunk 4]
```diff
@@ -157,6 +157,7 @@ def start_cloud_training(input_dict: Dict[str, List[types.Artifact]],
                          executor_class_path: str,
                          job_args: Dict[str, Any],
                          job_id: Optional[str],
+                         job_labels: Optional[Dict[str, Any]] = None, # AI Platform only.
```
Comment (chensun) [inline tfx/extensions/google_cloud_ai_platform/runner.py:160]: Adding a new parameter in the middle would be a breaking change for those rely on positional arguments. Do we have any concern on this? 
/cc @1025KB

Comment (axeltidemann) [reply tfx/extensions/google_cloud_ai_platform/runner.py:160]: I added it there because it semantically would make more sense to have `job_args, job_id, job_labels` grouped together. It can be moved to the end, of course. Or could this be listed in the `RELEASE.md` under `Breaking Changes`?

Comment (chensun) [reply tfx/extensions/google_cloud_ai_platform/runner.py:160]: Sounds good then. Thanks!
